### ISSUE TITLE
"error: Sequence contains no matching element" when listing outdated packages

### ISSUE THREAD

**Issue by strout** [NuGet/Home#12256]
### NuGet Product Used

dotnet.exe

### Product Version

7.0.100, 6.0.300, 5.0.408, 3.1.425 (all I tested)

### Worked before?

no

### Impact

It's more difficult to complete my work

### Repro Steps & Context

1. `dotnet new console -o repro`
2. `dotnet add repro package AngleSharp --version 1.0.0-alpha-844` (note: this version was unlisted, the highest still-published version is 0.17.1)
3. (optional) `dotnet add repro package AngleSharp.Css --version 0.16.0` (note: highest released version is 0.16.4)
4. `dotnet list repro package --outdated --highest-minor` (or `--highest-patch`; the equivalent `dotnet nuget` command also fails)

What I expected:

I would expect the command to complete successfully. At a minimum it should still list the other package that's outdated (AngleSharp.Css 0.16.0 -> 0.16.4) if step 3 was run. It would also be nice to have a warning [like the one on nuget.org](https://www.nuget.org/packages/AngleSharp/1.0.0-alpha-844) about the AngleSharp version being unlisted, with a mention of the highest published version.

What I got:

```
The following sources were used:
   https://api.nuget.org/v3/index.json

error: Sequence contains no matching element


Usage: NuGet.CommandLine.XPlat.dll package list [arguments] [options]
[...]
```

---

This made it more difficult to identify packages that needed I to update (note that `dotnet list repro package --outdated` also does not list AngleSharp) on a project with several dependencies. I hope it can be fixed to save others the same difficulty. Thanks for considering this issue.

### Verbose Logs

```shell
The following sources were used:
   https://api.nuget.org/v3/index.json

  CACHE https://api.nuget.org/v3/registration5-gz-semver2/anglesharp/index.json
  CACHE https://api.nuget.org/v3/registration5-gz-semver2/anglesharp/page/0.2.0/0.13.0-alpha-744.json
  CACHE https://api.nuget.org/v3/registration5-gz-semver2/anglesharp/page/0.13.0-alpha-745/0.16.1-alpha-167.json
  CACHE https://api.nuget.org/v3/registration5-gz-semver2/anglesharp/page/0.16.1-alpha-168/1.0.0-alpha-844.json
  CACHE https://api.nuget.org/v3/registration5-gz-semver2/anglesharp.css/index.json
  CACHE https://api.nuget.org/v3/registration5-gz-semver2/anglesharp/index.json
  CACHE https://api.nuget.org/v3/registration5-gz-semver2/anglesharp/page/0.2.0/0.13.0-alpha-744.json
  CACHE https://api.nuget.org/v3/registration5-gz-semver2/anglesharp/page/0.13.0-alpha-745/0.16.1-alpha-167.json
  CACHE https://api.nuget.org/v3/registration5-gz-semver2/anglesharp/page/0.16.1-alpha-168/1.0.0-alpha-844.json
  CACHE https://api.nuget.org/v3/registration5-gz-semver2/anglesharp.css/index.json
error: Sequence contains no matching element
trace: System.AggregateException: One or more errors occurred. (Sequence contains no matching element)
trace:  ---> System.InvalidOperationException: Sequence contains no matching element
trace:    at System.Linq.ThrowHelper.ThrowNoMatchException()
trace:    at System.Linq.Enumerable.First[TSource](IEnumerable`1 source, Func`2 predicate)
trace:    at NuGet.CommandLine.XPlat.ListPackageCommandRunner.GetVersionsFromDictAsync(IEnumerable`1 packages, Dictionary`2 packagesVersionsDict, ListPackageArgs listPackageArgs)
trace:    at NuGet.CommandLine.XPlat.ListPackageCommandRunner.AddLatestVersionsAsync(IEnumerable`1 packages, ListPackageArgs listPackageArgs)
trace:    at NuGet.CommandLine.XPlat.ListPackageCommandRunner.ExecuteCommandAsync(ListPackageArgs listPackageArgs)
trace:    at NuGet.CommandLine.XPlat.ListPackageCommand.<>c__DisplayClass0_1.<<Register>b__1>d.MoveNext()
trace:    --- End of inner exception stack trace ---
trace:    at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
trace:    at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
trace:    at NuGet.CommandLine.XPlat.Program.MainInternal(String[] args, CommandOutputLogger log)


Usage: NuGet.CommandLine.XPlat.dll package list [arguments] [options]

Arguments:
  <PROJECT | SOLUTION>  A path to a project, solution file or directory.

Options:
  -h|--help               Show help information
  --force-english-output  Forces the application to run using an invariant, English-based culture.
  --framework             Specifies the target framework for which the packages will be listed.
  --deprecated            Displays only the packages marked deprecated by the authors.
  --outdated              Displays only the packages that need updates with the latest version from the sources.
  --vulnerable            Displays only the packages flagged as vulnerable.
  --include-transitive    Includes transitive packages too in the result.
  --include-prerelease    Considers prerelease versions when looking for latest. Works only with `--outdated`.
  --highest-patch         Considers only the versions with matching minor and major. Works only with `--outdated`.
  --highest-minor         Considers only the versions with matching major. Works only with `--outdated`.
  --source                Sources to lookup for latest versions. Works only with `--outdated`.
  --config                A path to a config file to specify sources. Works only with `--outdated`.
  --interactive           Allow the command to block and require manual action for operations like authentication.
  -v|--verbosity          Set the verbosity level of the command. Allowed values are q[uiet], m[inimal], n[ormal], d[etailed], and diag[nostic].
```

**Comment by erdembayar** [comment]
I can repro this issue on sdk `6.0.306`, but it doesn't repro on sdk `6.0.111`. Looks like regression introduced between them.
I did make [change into this code](https://github.com/NuGet/NuGet.Client/commit/035850255a15b60437d22f9178c4206bafe0b6a9) path 2 weeks ago, but it's slated go to next sdk release, so I don't think that change caused this regression. 
As of today, latest sdk is `SDK 6.0.403`, but regression was in v`6.0.306`.

**Comment by erdembayar** [comment]
Fyi @nkolev92

**Comment by telb99** [comment]
Still an issue.
We updated StyleCop.Analyzers to v1.2.0-beta.507 recently and now getting this error.

=====

dotnet list "path_to.csproj" package --outdated --highest-patch

error: Sequence contains no matching element

Usage: NuGet.CommandLine.XPlat.dll package list [arguments] [options]

Arguments:
  <PROJECT | SOLUTION>  A path to a project, solution file or directory.

.NET SDK (7.0.306)
NuGet Version: 6.4.0.123
and 
NuGet Version: 6.6.1.2

Looks like code similar to this:
var matchingPackage = packageMetadata.Where(p => p.Key.Equals(topLevelPackage.Name, StringComparison.OrdinalIgnoreCase)).First();

is failing if metadata does not exist for package...
I suggest changing .First() to FirstOrDefault() and doing a "continue" if that returns default.

I was able to get around my problem by adding --include-prerelease and now have to deal with it including pre release packages all the time when generally I don't want them... We needed the prerelease version of StyleCop.Analyzers because while it is getting regular updates, it hasn't been released since 2019 and doesn't look like it will any time soon because new versions of .NET are coming out faster than it can be updated and regression tested.
